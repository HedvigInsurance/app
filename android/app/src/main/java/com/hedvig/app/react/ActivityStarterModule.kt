package com.hedvig.app.react

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.support.v4.app.FragmentActivity
import android.support.v4.app.FragmentManager
import android.support.v4.content.LocalBroadcastManager
import androidx.navigation.Navigation
import com.apollographql.apollo.ApolloClient
import com.apollographql.apollo.rx2.Rx2Apollo
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import com.google.firebase.analytics.FirebaseAnalytics
import com.hedvig.android.owldroid.graphql.InsuranceStatusQuery
import com.hedvig.android.owldroid.type.InsuranceStatus
import com.hedvig.app.LoggedInActivity
import com.hedvig.app.R
import com.hedvig.app.feature.chat.UploadBottomSheet
import com.hedvig.app.feature.dashboard.ui.PerilBottomSheet
import com.hedvig.app.feature.dashboard.ui.PerilIcon
import com.hedvig.app.feature.offer.OfferActivity
import com.hedvig.app.feature.offer.OfferChatOverlayFragment
import com.hedvig.app.feature.referrals.PromotionCodeBottomSheet
import com.hedvig.app.util.extensions.proxyNavigate
import com.hedvig.app.util.extensions.setIsLoggedIn
import com.hedvig.app.util.extensions.triggerRestartActivity
import com.hedvig.app.util.react.AsyncStorageNative
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.plusAssign
import timber.log.Timber

class ActivityStarterModule(
    reactContext: ReactApplicationContext,
    private val apolloClient: ApolloClient,
    private val asyncStorageNative: AsyncStorageNative
) : ReactContextBaseJavaModule(reactContext), LifecycleEventListener {

    private val disposables = CompositeDisposable()

    private val fragmentManager: FragmentManager
        get() = (reactApplicationContext.currentActivity as FragmentActivity).supportFragmentManager

    private val firebaseAnalytics by lazy { FirebaseAnalytics.getInstance(reactApplicationContext) }

    private val localBroadcastManager = LocalBroadcastManager.getInstance(reactContext)

    private val fileUploadBroadcastReceiver = FileUploadBroadcastReceiver()

    private var fileUploadCallback: Promise? = null

    init {
        localBroadcastManager.registerReceiver(fileUploadBroadcastReceiver, IntentFilter(FILE_UPLOAD_INTENT))
    }

    override fun getName() = "ActivityStarter"

    override fun onHostResume() {
    }

    override fun onHostPause() {
    }

    override fun onHostDestroy() {
        localBroadcastManager.unregisterReceiver(fileUploadBroadcastReceiver)
        disposables.clear()
    }

    @ReactMethod
    fun navigateToOfferFromChat() {
        val activity = reactApplicationContext.currentActivity
        if (activity != null) {
            asyncStorageNative.setKey("@hedvig:isViewingOffer", "true")
            currentActivity?.let {
                it.startActivity(Intent(it, OfferActivity::class.java))
            }
        }
    }

    @ReactMethod
    fun navigateToChatFromOffer() {
        currentActivity?.let {
            it.startActivity(Intent(it, OfferActivity::class.java))
        }
    }

    @ReactMethod
    fun navigateToLoggedInFromChat() {
        currentActivity?.let { activity ->
            reactApplicationContext.setIsLoggedIn(true)
            val intent = Intent(activity, LoggedInActivity::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
            activity.startActivity(intent)
        }
    }

    @ReactMethod
    fun showOfferChatOverlay() {
        val offerChatOverlayFragment = OfferChatOverlayFragment()
        offerChatOverlayFragment.show(fragmentManager, "OfferChatOverlay")
    }

    @ReactMethod
    fun showPerilOverlay(subject: String, iconId: String, title: String, description: String) {
        PerilBottomSheet.newInstance(subject, PerilIcon.from(iconId), title, description)
            .show(fragmentManager, "perilSheet")
    }

    @ReactMethod
    fun showPromotionOverlay() {
        PromotionCodeBottomSheet.newInstance()
            .show(fragmentManager, "perilSheet")
    }

    @ReactMethod
    fun showFileUploadOverlay(onUpload: Promise) {
        val uploadBottomSheet = UploadBottomSheet()
        uploadBottomSheet.show(fragmentManager, "FileUploadOverlay")
        fileUploadCallback = onUpload
    }

    @ReactMethod
    fun logEvent(eventName: String, map: ReadableMap) =
        firebaseAnalytics.logEvent(eventName, convertMapToBundle(map))

    @ReactMethod
    fun restartApplication() =
        reactApplicationContext.currentActivity?.let {
            it.triggerRestartActivity(it::class.java)
        }

    @ReactMethod
    fun reloadChat() =
        localBroadcastManager.sendBroadcast(Intent(BROADCAST_RELOAD_CHAT))

    @ReactMethod
    fun doIsLoggedInProcedure() {
        disposables += Rx2Apollo
            .from(apolloClient.query(InsuranceStatusQuery()))
            .subscribe({ response ->
                response.data()?.insurance?.status?.let { status ->
                    when (status) {
                        InsuranceStatus.INACTIVE,
                        InsuranceStatus.INACTIVE_WITH_START_DATE,
                        InsuranceStatus.ACTIVE -> navigateToLoggedInFromChat()
                        InsuranceStatus.PENDING,
                        InsuranceStatus.TERMINATED,
                        InsuranceStatus.`$UNKNOWN` -> {
                            // Let the chat take care of this
                        }
                    }
                }
            }, { Timber.e(it) })
    }

    private fun convertMapToBundle(readableMap: ReadableMap): Bundle {
        val bundle = Bundle()
        val iterator = readableMap.keySetIterator()
        while (iterator.hasNextKey()) {
            val key = iterator.nextKey()
            when (readableMap.getType(key)) {
                ReadableType.Boolean -> bundle.putBoolean(key, readableMap.getBoolean(key))
                ReadableType.Number -> bundle.putDouble(key, readableMap.getDouble(key))
                ReadableType.String -> bundle.putString(key, readableMap.getString(key))
                ReadableType.Map -> bundle.putBundle(key, convertMapToBundle(readableMap.getMap(key)))
                else -> {
                    Timber.e("We don't cover null or array")
                }
            }
        }
        return bundle
    }

    private inner class FileUploadBroadcastReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            when (intent.getStringExtra(FILE_UPLOAD_RESULT)) {
                FILE_UPLOAD_SUCCESS -> {
                    fileUploadCallback?.resolve(intent.getStringExtra(FILE_UPLOAD_KEY))
                        ?: Timber.e("File upload successful but no callback present")
                    fileUploadCallback = null
                }
                FILE_UPLOAD_ERROR -> {
                    fileUploadCallback?.reject("E_NETWORK_ERROR", "failed to upload")
                        ?: Timber.e("File upload failed but no callback present") // TODO improve
                    fileUploadCallback = null
                }
            }
        }
    }

    companion object {
        const val BROADCAST_RELOAD_CHAT = "reloadChat"
        const val REDEEMED_CODE_BROADCAST = "redeemedCode"

        const val BROADCAST_MESSAGE_NAME = "message"

        const val MESSAGE_PROMOTION_CODE_REDEEMED = "promotionCodeRedeemed"

        const val FILE_UPLOAD_INTENT = "file_upload"
        const val FILE_UPLOAD_RESULT = "file_upload_result"
        const val FILE_UPLOAD_SUCCESS = "success"
        const val FILE_UPLOAD_ERROR = "error"
        const val FILE_UPLOAD_KEY = "key"
    }
}
