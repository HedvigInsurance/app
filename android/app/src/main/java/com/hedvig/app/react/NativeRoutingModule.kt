package com.hedvig.app.react

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.support.v4.content.LocalBroadcastManager
import com.apollographql.apollo.ApolloClient
import com.apollographql.apollo.rx2.Rx2Apollo
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.firestore.FirebaseFirestore
import com.hedvig.android.owldroid.graphql.MemberIdQuery
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.plusAssign
import timber.log.Timber
import java.util.Date
import java.util.HashMap

class NativeRoutingModule constructor(
    reactContext: ReactApplicationContext,
    private val apolloClient: ApolloClient
) : ReactContextBaseJavaModule(reactContext), LifecycleEventListener {

    private val profileBroadcastReceiver = ProfileBroadcastReceiver()
    private val onBoardingBroadcastReceiver = OnBoardingBroadcastReceiver()

    private val localBroadcastManager = LocalBroadcastManager.getInstance(reactContext)

    private val deviceEventEmitter: DeviceEventManagerModule.RCTDeviceEventEmitter
        get() = reactApplicationContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)

    private val compositeDisposable = CompositeDisposable()

    init {
        reactContext.addLifecycleEventListener(this)
    }

    override fun getName() = "NativeRouting"

    override fun onHostResume() {
        localBroadcastManager.registerReceiver(profileBroadcastReceiver, IntentFilter("profileNavigation"))
        localBroadcastManager.registerReceiver(onBoardingBroadcastReceiver, IntentFilter(ON_BOARDING_INTENT_FILER))
    }

    override fun onHostPause() {
        localBroadcastManager.unregisterReceiver(profileBroadcastReceiver)
        localBroadcastManager.unregisterReceiver(onBoardingBroadcastReceiver)
    }

    override fun onHostDestroy() {
    }

    @ReactMethod
    fun userDidSign() {
        val referralsStorage = currentActivity!!.getSharedPreferences("referrals", Context.MODE_PRIVATE)
        val referee = referralsStorage
            .getString("referee", null)
        val incentive = referralsStorage
            .getString("incentive", null)

        if (referee == null || incentive == null) {
            return
        }

        val memberIdQuery = MemberIdQuery()
        compositeDisposable +=
            Rx2Apollo.from(apolloClient.query(memberIdQuery)).subscribe({ response ->
                val data = response.data()
                if (data == null) {
                    Timber.e("Data was null when trying to accredit referral")
                    return@subscribe
                }
                val member = data.member()
                val memberId = member.id()
                if (memberId == null) {
                    Timber.e("memberId was null when trying to accredit referral")
                    return@subscribe
                }

                val b = Bundle()
                b.putString("invitedByMemberId", referee)
                b.putString("memberId", memberId)
                b.putString("incentive", incentive)

                FirebaseAnalytics
                    .getInstance(reactApplicationContext)
                    .logEvent("referrals_sign", b)

                val referralParameters = HashMap<String, String>()
                referralParameters["invitedByMemberId"] = referee
                referralParameters["memberId"] = memberId
                referralParameters["incentive"] = incentive
                referralParameters["timestamp"] = java.lang.Long.toString(Date().time)

                FirebaseFirestore
                    .getInstance()
                    .collection("referrals")
                    .add(referralParameters)
                    .addOnSuccessListener { Timber.i("Successfully saved referral!") }
                    .addOnFailureListener { error -> Timber.e(error, "Failed to save referral") }

            }, { error -> Timber.e(error, "Failed to load memberId for referral") })
    }

    private fun logoutAndRestartApplication() {
        deviceEventEmitter.emit("NativeRoutingLogoutAndRestartApplication", null)
    }

    private fun restartChatOnBoarding() {
        deviceEventEmitter.emit("NativeRoutingRestartChatOnBoarding", null)
    }

    private inner class ProfileBroadcastReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            when (intent.getStringExtra(NAVIGATE_ROUTING_EXTRA_NAME_ACTION)) {
                NAVIGATE_ROUTING_EXTRA_VALUE_LOGOUT_AND_RESTART_APPLICATION -> {
                    logoutAndRestartApplication()
                }
            }
        }
    }

    private inner class OnBoardingBroadcastReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            when (intent.getStringExtra(NAVIGATE_ROUTING_EXTRA_NAME_ACTION)) {
                NAVIGATE_ROUTING_EXTRA_VALUE_LOGOUT_AND_RESTART_APPLICATION -> logoutAndRestartApplication()
                NAVIGATE_ROUTING_EXTRA_VALUE_RESTART_CHAT_ON_BOARDING -> restartChatOnBoarding()
            }
        }
    }

    companion object {

        const val ON_BOARDING_INTENT_FILER = "on_boarding_intent_filter"

        const val NAVIGATE_ROUTING_EXTRA_NAME_ACTION = "action"

        const val NAVIGATE_ROUTING_EXTRA_VALUE_LOGOUT_AND_RESTART_APPLICATION = "logoutAndRestartApplication"
        const val NAVIGATE_ROUTING_EXTRA_VALUE_RESTART_CHAT_ON_BOARDING = "restartChatOnBoarding"
    }
}
