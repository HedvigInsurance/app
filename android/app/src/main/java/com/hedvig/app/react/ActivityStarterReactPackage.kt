package com.hedvig.app.react

import com.apollographql.apollo.ApolloClient
import com.facebook.react.ReactPackage
import com.facebook.react.bridge.NativeModule
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.uimanager.ViewManager
import com.hedvig.app.util.react.AsyncStorageNative
import java.util.ArrayList

class ActivityStarterReactPackage(
    private val apolloClient: ApolloClient,
    private val asyncStorageNative: AsyncStorageNative
) : ReactPackage {

    override fun createNativeModules(
        reactContext: ReactApplicationContext
    ): List<NativeModule> {
        val modules = ArrayList<NativeModule>()
        modules.add(ActivityStarterModule(reactContext, apolloClient, asyncStorageNative))
        return modules
    }

    override fun createViewManagers(reactContext: ReactApplicationContext): List<ViewManager<*, *>> {
        return emptyList()
    }
}
