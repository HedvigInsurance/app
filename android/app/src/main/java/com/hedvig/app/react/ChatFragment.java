package com.hedvig.app.react;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.facebook.react.ReactInstanceManager;
import com.facebook.react.ReactRootView;
import com.facebook.react.modules.core.DefaultHardwareBackBtnHandler;
import dagger.android.support.AndroidSupportInjection;
import javax.inject.Inject;

public class ChatFragment extends Fragment implements DefaultHardwareBackBtnHandler {

    @Inject
    ReactInstanceManager reactInstanceManager;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        AndroidSupportInjection.inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        ReactRootView reactRootView = new ReactRootView(requireContext());
        reactRootView.startReactApplication(reactInstanceManager, "Chat", null);
        return reactRootView;
    }

    @Override
    public void invokeDefaultOnBackPressed() {
        if (getActivity() != null) {
            getActivity().onBackPressed();
        }
    }
}
