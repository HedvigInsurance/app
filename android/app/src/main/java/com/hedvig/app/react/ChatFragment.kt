package com.hedvig.app.react

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.facebook.react.ReactApplication
import com.facebook.react.ReactInstanceManager
import com.facebook.react.ReactNativeHost
import com.facebook.react.ReactRootView
import com.facebook.react.common.LifecycleState
import com.facebook.react.modules.core.DefaultHardwareBackBtnHandler

import dagger.android.support.AndroidSupportInjection

class ChatFragment : Fragment(), DefaultHardwareBackBtnHandler {

    private var mReactRootView: ReactRootView? = null

    protected val reactNativeHost: ReactNativeHost
        get() = (activity!!.application as ReactApplication).reactNativeHost

    internal val reactInstanceManager: ReactInstanceManager
        get() = reactNativeHost.reactInstanceManager

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        AndroidSupportInjection.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val reactRootView = ReactRootView(requireContext())
        mReactRootView = reactRootView
        reactRootView.startReactApplication(reactInstanceManager, "Chat", arguments)
        return reactRootView
    }

    override fun onDestroy() {
        super.onDestroy()
        if (mReactRootView != null) {
            mReactRootView!!.unmountReactApplication()
            mReactRootView = null
        }
        if (reactInstanceManager.lifecycleState != LifecycleState.RESUMED) {
            reactInstanceManager.onHostDestroy(activity)
            reactNativeHost.clear()
        }
    }

    override fun invokeDefaultOnBackPressed() {
        if (activity != null) {
            activity!!.onBackPressed()
        }
    }
}
