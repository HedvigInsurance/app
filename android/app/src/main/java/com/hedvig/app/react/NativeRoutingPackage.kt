package com.hedvig.app.react

import com.apollographql.apollo.ApolloClient
import com.facebook.react.ReactPackage
import com.facebook.react.bridge.NativeModule
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.uimanager.ViewManager

class NativeRoutingPackage constructor(private val apolloClient: ApolloClient) : ReactPackage {

    override fun createNativeModules(reactContext: ReactApplicationContext): List<NativeModule> =
        listOf(NativeRoutingModule(reactContext, apolloClient))

    override fun createViewManagers(reactContext: ReactApplicationContext): List<ViewManager<*, *>> = emptyList()
}
