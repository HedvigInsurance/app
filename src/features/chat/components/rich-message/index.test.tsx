import * as React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { RichMessage } from './';

test('renders giphy messages correctly', () => {
  const component = mount(
    <RichMessage
      withMargin={false}
      message={{
        body: { text: 'https://giphy.com/mock.gif' },
        header: {
          editAllowed: false,
          statusMessage: 'mock',
        },
      }}
      index={0}
    />,
  );

  expect(toJson(component)).toMatchSnapshot();
});

test('renders image messages correctly', () => {
  const component = mount(
    <RichMessage
      withMargin={false}
      message={{
        body: { text: 'https://randomwebsite.com/image.jpg' },
        header: {
          editAllowed: false,
          statusMessage: 'mock',
        },
      }}
      index={0}
    />,
  );

  expect(toJson(component)).toMatchSnapshot();
});

test('renders text messages correctly', () => {
  const component = mount(
    <RichMessage
      withMargin={false}
      message={{
        body: { text: 'mock message hello 123' },
        header: {
          editAllowed: false,
          statusMessage: 'mock',
        },
      }}
      index={0}
    />,
  );

  expect(toJson(component)).toMatchSnapshot();
});

test('renders text messages with hyperlinks correctly', () => {
  const component = mount(
    <RichMessage
      withMargin={false}
      message={{
        body: { text: 'mock message hello 123 https://www.hedvig.com' },
        header: {
          editAllowed: false,
          statusMessage: 'mock',
        },
      }}
      index={0}
    />,
  );

  expect(toJson(component)).toMatchSnapshot();
});
