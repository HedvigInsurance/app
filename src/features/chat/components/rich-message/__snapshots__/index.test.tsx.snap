// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders giphy messages correctly 1`] = `
<Component
  fromUser={true}
  index={0}
  message={
    Object {
      "body": Object {
        "text": "https://giphy.com/mock.gif",
      },
      "header": Object {
        "editAllowed": false,
        "statusMessage": "mock",
      },
    }
  }
  withMargin={false}
>
  <Component
    fromUser={true}
    index={0}
    message={
      Object {
        "body": Object {
          "text": "https://giphy.com/mock.gif",
        },
        "header": Object {
          "editAllowed": false,
          "statusMessage": "mock",
        },
      }
    }
    withMargin={false}
  >
    <emotion(View)
      withMargin={false}
    >
      <View
        style={
          Object {
            "borderRadius": 20,
            "marginBottom": 0,
            "overflow": "hidden",
          }
        }
      >
        <react-native-mock>
          <emotion(ImageProgress)
            indicator={[Function]}
            indicatorProps={
              Object {
                "color": "#FF8A80",
                "size": 40,
                "thickness": 5,
              }
            }
            source={
              Object {
                "uri": "https://giphy.com/mock.gif",
              }
            }
          >
            <ImageProgress
              errorContainerStyle={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
              indicator={[Function]}
              indicatorContainerStyle={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
              indicatorProps={
                Object {
                  "color": "#FF8A80",
                  "size": 40,
                  "thickness": 5,
                }
              }
              source={
                Object {
                  "uri": "https://giphy.com/mock.gif",
                }
              }
              style={
                Object {
                  "height": 200,
                  "width": 280,
                }
              }
              threshold={50}
            >
              <View
                style={
                  Object {
                    "height": 200,
                    "width": 280,
                  }
                }
              >
                <react-native-mock>
                  <Image
                    key="https://giphy.com/mock.gif"
                    onError={[Function]}
                    onLoad={[Function]}
                    onLoadEnd={[Function]}
                    onLoadStart={[Function]}
                    onProgress={[Function]}
                    source={
                      Object {
                        "uri": "https://giphy.com/mock.gif",
                      }
                    }
                    style={
                      Array [
                        undefined,
                        undefined,
                      ]
                    }
                  >
                    <react-native-mock />
                  </Image>
                </react-native-mock>
              </View>
            </ImageProgress>
          </emotion(ImageProgress)>
          <emotion(View)>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "borderBottomLeftRadius": 20,
                  "borderBottomRightRadius": 20,
                  "borderColor": "#E9ECEF",
                  "borderTopWidth": 0,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "padding": 10,
                  "paddingLeft": 15,
                }
              }
            >
              <react-native-mock>
                <Component
                  height={20}
                  width={20}
                >
                  <Svg
                    height={20}
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 38 48"
                    width={20}
                  >
                    <RNSVGSvgView
                      align="xMidYMid"
                      bbHeight="20"
                      bbWidth="20"
                      meetOrSlice={0}
                      minX={0}
                      minY={0}
                      style={
                        Array [
                          Object {
                            "backgroundColor": "transparent",
                          },
                          undefined,
                          false,
                          Object {
                            "flex": 0,
                            "height": 20,
                            "width": 20,
                          },
                        ]
                      }
                      vbHeight={48}
                      vbWidth={38}
                    />
                  </Svg>
                </Component>
                <emotion(View)
                  width={10}
                >
                  <View
                    style={
                      Object {
                        "height": 1,
                        "width": 10,
                      }
                    }
                    width={10}
                  >
                    <react-native-mock />
                  </View>
                </emotion(View)>
                <emotion(Text)>
                  <Text
                    style={
                      Object {
                        "color": "#9B9BAA",
                        "fontFamily": "Helvetica Neue",
                        "fontSize": 12,
                        "fontWeight": "500",
                        "marginTop": 1,
                      }
                    }
                  >
                    <react-native-mock>
                      GIPHY
                    </react-native-mock>
                  </Text>
                </emotion(Text)>
              </react-native-mock>
            </View>
          </emotion(View)>
        </react-native-mock>
      </View>
    </emotion(View)>
  </Component>
</Component>
`;

exports[`renders image messages correctly 1`] = `
<Component
  fromUser={true}
  index={0}
  message={
    Object {
      "body": Object {
        "text": "https://randomwebsite.com/image.jpg",
      },
      "header": Object {
        "editAllowed": false,
        "statusMessage": "mock",
      },
    }
  }
  withMargin={false}
>
  <Component
    fromUser={true}
    index={0}
    message={
      Object {
        "body": Object {
          "text": "https://randomwebsite.com/image.jpg",
        },
        "header": Object {
          "editAllowed": false,
          "statusMessage": "mock",
        },
      }
    }
    withMargin={false}
  >
    <emotion(View)
      withMargin={false}
    >
      <View
        style={
          Object {
            "borderColor": "#E9ECEF",
            "borderRadius": 20,
            "borderWidth": undefined,
            "marginBottom": 0,
            "overflow": "hidden",
          }
        }
      >
        <react-native-mock>
          <emotion(ImageProgress)
            indicator={[Function]}
            indicatorProps={
              Object {
                "color": "#651EFF",
                "size": 40,
                "thickness": 3,
              }
            }
            source={
              Object {
                "uri": "undefined/unsafe/400x560/smart/https%3A%2F%2Frandomwebsite.com%2Fimage.jpg",
              }
            }
          >
            <ImageProgress
              errorContainerStyle={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
              indicator={[Function]}
              indicatorContainerStyle={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
              indicatorProps={
                Object {
                  "color": "#651EFF",
                  "size": 40,
                  "thickness": 3,
                }
              }
              source={
                Object {
                  "uri": "undefined/unsafe/400x560/smart/https%3A%2F%2Frandomwebsite.com%2Fimage.jpg",
                }
              }
              style={
                Object {
                  "height": 200,
                  "width": 280,
                }
              }
              threshold={50}
            >
              <View
                style={
                  Object {
                    "height": 200,
                    "width": 280,
                  }
                }
              >
                <react-native-mock>
                  <Image
                    key="undefined/unsafe/400x560/smart/https%3A%2F%2Frandomwebsite.com%2Fimage.jpg"
                    onError={[Function]}
                    onLoad={[Function]}
                    onLoadEnd={[Function]}
                    onLoadStart={[Function]}
                    onProgress={[Function]}
                    source={
                      Object {
                        "uri": "undefined/unsafe/400x560/smart/https%3A%2F%2Frandomwebsite.com%2Fimage.jpg",
                      }
                    }
                    style={
                      Array [
                        undefined,
                        undefined,
                      ]
                    }
                  >
                    <react-native-mock />
                  </Image>
                </react-native-mock>
              </View>
            </ImageProgress>
          </emotion(ImageProgress)>
        </react-native-mock>
      </View>
    </emotion(View)>
  </Component>
</Component>
`;

exports[`renders text messages correctly 1`] = `
<Component
  fromUser={true}
  index={0}
  message={
    Object {
      "body": Object {
        "text": "mock message hello 123",
      },
      "header": Object {
        "editAllowed": false,
        "statusMessage": "mock",
      },
    }
  }
  withMargin={false}
>
  <Component
    fromUser={true}
    index={0}
    message={
      Object {
        "body": Object {
          "text": "mock message hello 123",
        },
        "header": Object {
          "editAllowed": false,
          "statusMessage": "mock",
        },
      }
    }
    withMargin={false}
  >
    <Container
      actions={
        Object {
          "setShowEditDialog": [Function],
        }
      }
      initialState={
        Object {
          "showEditDialog": false,
        }
      }
    >
      <InnerContainer
        actions={
          Object {
            "setShowEditDialog": [Function],
          }
        }
        initialState={
          Object {
            "showEditDialog": false,
          }
        }
      >
        <StyledUserChatMessage
          withMargin={false}
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "backgroundColor": "#651EFF",
                  "borderColor": "#651EFF",
                  "borderRadius": 24,
                  "borderWidth": 1,
                  "justifyContent": "center",
                  "marginBottom": 2,
                  "paddingBottom": 8,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "paddingTop": 8,
                },
                undefined,
              ]
            }
          >
            <react-native-mock>
              <_class
                linkDefault={true}
                linkStyle={
                  Object {
                    "textDecorationLine": "underline",
                  }
                }
              >
                <Hyperlink
                  linkDefault={true}
                  linkStyle={
                    Object {
                      "textDecorationLine": "underline",
                    }
                  }
                  onPress={[Function]}
                >
                  <View>
                    <react-native-mock>
                      <StyledDefaultUserMessageText
                        fromUser={true}
                        key=".0"
                      >
                        <Text
                          fromUser={true}
                          selectable={true}
                          style={
                            Array [
                              undefined,
                              Object {
                                "color": "#FFFFFF",
                                "fontFamily": "CircularStd-Book",
                                "fontSize": 16,
                                "textAlign": "left",
                              },
                            ]
                          }
                        >
                          <react-native-mock>
                            mock message hello 123
                          </react-native-mock>
                        </Text>
                      </StyledDefaultUserMessageText>
                    </react-native-mock>
                  </View>
                </Hyperlink>
              </_class>
            </react-native-mock>
          </View>
        </StyledUserChatMessage>
      </InnerContainer>
    </Container>
  </Component>
</Component>
`;

exports[`renders text messages with hyperlinks correctly 1`] = `
<Component
  fromUser={true}
  index={0}
  message={
    Object {
      "body": Object {
        "text": "mock message hello 123 https://www.hedvig.com",
      },
      "header": Object {
        "editAllowed": false,
        "statusMessage": "mock",
      },
    }
  }
  withMargin={false}
>
  <Component
    fromUser={true}
    index={0}
    message={
      Object {
        "body": Object {
          "text": "mock message hello 123 https://www.hedvig.com",
        },
        "header": Object {
          "editAllowed": false,
          "statusMessage": "mock",
        },
      }
    }
    withMargin={false}
  >
    <Container
      actions={
        Object {
          "setShowEditDialog": [Function],
        }
      }
      initialState={
        Object {
          "showEditDialog": false,
        }
      }
    >
      <InnerContainer
        actions={
          Object {
            "setShowEditDialog": [Function],
          }
        }
        initialState={
          Object {
            "showEditDialog": false,
          }
        }
      >
        <StyledUserChatMessage
          withMargin={false}
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "backgroundColor": "#651EFF",
                  "borderColor": "#651EFF",
                  "borderRadius": 24,
                  "borderWidth": 1,
                  "justifyContent": "center",
                  "marginBottom": 2,
                  "paddingBottom": 8,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "paddingTop": 8,
                },
                undefined,
              ]
            }
          >
            <react-native-mock>
              <_class
                linkDefault={true}
                linkStyle={
                  Object {
                    "textDecorationLine": "underline",
                  }
                }
              >
                <Hyperlink
                  linkDefault={true}
                  linkStyle={
                    Object {
                      "textDecorationLine": "underline",
                    }
                  }
                  onPress={[Function]}
                >
                  <View>
                    <react-native-mock>
                      <StyledDefaultUserMessageText
                        fromUser={true}
                        key=".0"
                      >
                        <Text
                          fromUser={true}
                          selectable={true}
                          style={
                            Array [
                              undefined,
                              Object {
                                "color": "#FFFFFF",
                                "fontFamily": "CircularStd-Book",
                                "fontSize": 16,
                                "textAlign": "left",
                              },
                            ]
                          }
                        >
                          <react-native-mock>
                            <Text
                              fromUser={true}
                              key=".0"
                            >
                              <react-native-mock>
                                mock message hello 123 
                                <Text
                                  fromUser={true}
                                  key="https://www.hedvig.com23"
                                  onPress={[Function]}
                                  style={
                                    Array [
                                      undefined,
                                      Object {
                                        "textDecorationLine": "underline",
                                      },
                                    ]
                                  }
                                >
                                  <react-native-mock>
                                    https://www.hedvig.com
                                  </react-native-mock>
                                </Text>
                              </react-native-mock>
                            </Text>
                          </react-native-mock>
                        </Text>
                      </StyledDefaultUserMessageText>
                    </react-native-mock>
                  </View>
                </Hyperlink>
              </_class>
            </react-native-mock>
          </View>
        </StyledUserChatMessage>
      </InnerContainer>
    </Container>
  </Component>
</Component>
`;
